datamade/bga-payroll
# Create entity metadata API

## Description

This PR installs [Django REST framework](https://www.django-rest-framework.org/#quickstart) and migrates calculations for display from view code to API endpoints for the index, units, and departments, and people.

The API expects a data year query parameter, however with the exception of the index endpoint and source file links, it does not yet use data year to filter calculations. This is so we can confirm the endpoints replicate the calculations from the views, before we add additional functionality in a separate PR.

Most of the diff in this PR is reorganizing code!

- All of the employer calculations are migrated verbatim from the view code, i.e., they don't need additional review. The only exception is grabbing salaries for an employer, which was migrated from the Salary model to a function on the Employer model in https://github.com/datamade/bga-payroll/commit/369c5dee3359d557513e65c42483bea4f8697750.
- Most of the person calculations were also migrated verbatim, except that the API returns total pay for current salary, instead of base pay.

## Test instructions

- Shut down any running containers: `docker-compose down`
- Build the application container to install the new dependency: `docker-compose build app`.
- Run the app: `docker-compose up -d`.
- Assuming your local instance has data, navigate to any unit page. Copy the slug from the URL.
- In a separate tab, navigate to `http://localhost:8000/units/SLUG/?data_year=2018`. Confirm that the data is returned without error, and that it matches what is displayed on the unit page.
- Repeat the above two steps for a department and a person, using the following endpoints:
  - `http://localhost:8000/departments/SLUG/?data_year=2018`
  - `http://localhost:8000/people/SLUG/?data_year=2018`
    - Note that `current_salary` is base pay in the view, but total pay in the API.
